Libplanet/Store/Trie/MerkleTrie.cs

```diff
@@ -0,0 +1,333 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Security.Cryptography;
+using Bencodex;
+using Bencodex.Types;
+using Libplanet.Store.Trie.Nodes;
+
+namespace Libplanet.Store.Trie
+{
+    // TODO: implement 'logs' for debugging.
+    [Equals]
+    internal class MerkleTrie : ITrie
+    {
+        private static Codec _codec;
+
+        private readonly bool _secure;
+
+        static MerkleTrie()
+        {
+            _codec = new Codec();
+        }
+
+        /// <summary>
+        /// An <see cref="ITrie"/> implementation.
+        /// </summary>
+        /// <param name="keyValueStore">The <see cref="IKeyValueStore"/> storage to store
+        /// nodes.</param>
+        /// <param name="root">The root node of <see cref="MerkleTrie"/>. If it is <c>null</c>,
+        /// it will be treated like empty trie.</param>
+        /// <param name="secure">The flag to determine to use <see cref="MerkleTrie"/> in secure
+        /// mode. If it is true, <see cref="MerkleTrie"/> will stores the value with the hashed
+        /// result from the given key as the key. It will hash with
+        /// <see cref="Hashcash.Hash"/>.</param>
+        public MerkleTrie(IKeyValueStore keyValueStore, INode root = null, bool secure = false)
+        {
+            KeyValueStore = keyValueStore;
+            Root = root;
+            _secure = secure;
+        }
+
+        /// <inheritdoc/>
+        public INode Root { get; private set; }
+
+        private IKeyValueStore KeyValueStore { get; }
+
+        public static bool operator ==(MerkleTrie left, MerkleTrie right) =>
+            Operator.Weave(left, right);
+
+        public static bool operator !=(MerkleTrie left, MerkleTrie right) =>
+            Operator.Weave(left, right);
+
+        /// <inheritdoc/>
+        public void Set(byte[] key, IValue value)
+        {
+            Root = Insert(
+                Root,
+                ImmutableArray<byte>.Empty,
+                ToKey(key).ToImmutableArray(),
+                new ValueNode(value));
+        }
+
+        /// <inheritdoc/>
+        public bool TryGet(byte[] key, out IValue value)
+        {
+            return TryGet(
+                Root,
+                ImmutableArray<byte>.Empty,
+                ToKey(key).ToImmutableArray(),
+                out value);
+        }
+
+        /// <inheritdoc/>
+        public ITrie Commit()
+        {
+            var newRoot = Commit(Root);
+
+            // It assumes embedded node if it's not HashNode.
+            if (!(newRoot is HashNode))
+            {
+                KeyValueStore.Set(newRoot.Hash().ToByteArray(), newRoot.Serialize());
+            }
+
+            return new MerkleTrie(KeyValueStore, newRoot);
+        }
+
+        private INode Commit(INode node)
+        {
+            switch (node)
+            {
+                case HashNode _:
+                    return node;
+
+                case FullNode fullNode:
```

Comment:
Since the cases are quite long, how about separating these with private methods like `CommitFullNode`, `CommitShortNode`, `CommitValueNode`?